ledger/complete/mtrie/trieCache.go

```diff
@@ -0,0 +1,149 @@
+package mtrie
+
+import (
+	"sync"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+)
+
+type OnTreeEvictedFunc func(tree *trie.MTrie)
+
+// TrieCache caches tries into memory, it acts as a fifo queue
+// so when it reaches to the capacity it would evict the oldest trie
+// from the cache.
+//
+// Under the hood it uses a circular buffer
+// of mtrie pointers and a map of rootHash to cache index for fast lookup
+type TrieCache struct {
+	tries         []*trie.MTrie
+	lookup        map[ledger.RootHash]int // index to item
+	lock          sync.RWMutex
+	capacity      int
+	tail          int // element index to write to
+	count         int // number of elements (count <= capacity)
+	onTreeEvicted OnTreeEvictedFunc
+}
+
+// NewTrieCache returns a new TrieCache with given capacity.
+func NewTrieCache(capacity uint, onTreeEvicted OnTreeEvictedFunc) *TrieCache {
+	return &TrieCache{
+		tries:         make([]*trie.MTrie, capacity),
+		lookup:        make(map[ledger.RootHash]int, capacity),
+		lock:          sync.RWMutex{},
+		capacity:      int(capacity),
+		tail:          0,
+		count:         0,
+		onTreeEvicted: onTreeEvicted,
+	}
+}
+
+// Purge removes all mtries stored in the buffer
+func (tc *TrieCache) Purge() {
+	tc.lock.Lock()
+	defer tc.lock.Unlock()
+
+	tc.tail = 0
+	tc.count = 0
+	for i := 0; i < tc.capacity; i++ {
+		if tc.onTreeEvicted != nil {
+			if tc.tries[i] != nil {
+				tc.onTreeEvicted(tc.tries[i])
+			}
+		}
+		tc.tries[i] = nil
+	}
+	tc.lookup = make(map[ledger.RootHash]int, tc.capacity)
+}
+
+// Tries returns elements in queue, starting from the oldest element
+// to the newest element.
+func (tc *TrieCache) Tries() []*trie.MTrie {
+	tc.lock.RLock()
+	defer tc.lock.RUnlock()
+
+	if tc.count == 0 {
+		return nil
+	}
+
+	tries := make([]*trie.MTrie, tc.count)
+
+	if tc.isFull() {
+		// If queue is full, tail points to the oldest element.
+		head := tc.tail
+		n := copy(tries, tc.tries[head:])
+		copy(tries[n:], tc.tries[:tc.tail])
+	} else {
+		if tc.tail >= tc.count { // Data isn't wrapped around the slice.
+			head := tc.tail - tc.count
+			copy(tries, tc.tries[head:tc.tail])
+		} else { // q.tail < q.count, data is wrapped around the slice.
+			// This branch isn't used until TrieQueue supports Pop (removing oldest element).
+			// At this time, there is no reason to implement Pop, so this branch is here to prevent future bug.
+			head := tc.capacity - tc.count + tc.tail
+			n := copy(tries, tc.tries[head:])
+			copy(tries[n:], tc.tries[:tc.tail])
+		}
+	}
+
+	return tries
+}
+
+// Push pushes trie to queue.  If queue is full, it overwrites the oldest element.
+func (tc *TrieCache) Push(t *trie.MTrie) {
+	tc.lock.Lock()
+	defer tc.lock.Unlock()
+
+	if tc.isFull() {
+		oldtrie := tc.tries[tc.tail]
+		if tc.onTreeEvicted != nil {
+			tc.onTreeEvicted(oldtrie)
```

Comment:
how heavy is this callback?  It will be run under the write lock now.

Reply:
Currently `onTreeEvicted` is nil, so it is a placeholder in case we need it.  

If callback is heavy, we can take a snapshot of tries using `Tries()` and call `onTreeEvicted` outside write lock. However, the tradeoff is `Tries()` would use heap alloc.

Reply:
if you're worry about running the callback while locking, just convert this into a helper method that return the evicted trie (or nil), and run the callback afterward.  e.g.,

```
func (tc *TrieCache) Push(t *trie.MTrie) {
  evicted := tc.push(t)
  if evicted != nil && tc.onTreeEvicted != nil {
    tc.onTreeEvicted(evicted)
  }
}

func (tc *TrieCache) push(t *trie.MTrie) *trie.MTrie {
   ... the original Push method ...
}
```

Reply:
yeah mostly kept for compatibility with upper layer, but in this version we pass nil for production